mycpp/util.py

```diff
@@ -110,3 +111,57 @@ def ShouldSkipPyFile(node: MypyFile) -> bool:
 def IsStr(t: Type):
     """Helper to check if a type is a string."""
     return isinstance(t, Instance) and t.type.fullname == 'builtins.str'
+
+
+def MaybeSkipIfStmt(visitor, stmt: IfStmt) -> bool:
+    """Returns true if the caller should not visit the entire if statement."""
+    cond = stmt.expr[0]
+
+    # Omit anything that looks like if __name__ == ...
+    if (isinstance(cond, ComparisonExpr) and
+            isinstance(cond.operands[0], NameExpr) and
+            cond.operands[0].name == '__name__'):
+        return True
+
+    if isinstance(cond, IntExpr) and cond.value == 0:
+        # But write else: body
+        # Note: this would be invalid at the top level!
+        if stmt.else_body:
+            visitor.accept(stmt.else_body)
+
+        return True
+
+    if isinstance(cond, NameExpr) and cond.name == 'TYPE_CHECKING':
+        # Omit if TYPE_CHECKING blocks.  They contain type expressions that
+        # don't type check!
+        return True
+
+    if isinstance(cond, MemberExpr) and cond.name == 'CPP':
+        # just take the if block
+        if hasattr(visitor, 'def_write_ind'):
```

Comment:
these `hasattr()` checks aren't the cutest, but hopefully the usage is obvious enough given the small number of users...

Reply:
Hmm yeah this seems weird ... because the new function doesn't just do a check if you should skip, it also visits.

Is it possible to 

1) share the first 3 if statements?
2) not share the def_write_ind() part?    


I remember having trouble with the visitor, so I won't be surprised if this is hard / annoying ...

But I don't see why not right now


----

Is it because def_write_ind() is in multiple passes?

Reply:
Yeah, it's pretty tricky. `def_rite_ind()` is only called in `cppgen_pass` IIRC.

Sharing the first 3 statements would still require this function to accept the visitor as an arg (so the `if 0:` case can visit the else block). If limit ourselves to the `TYPE_CHECKING` and` __name__` cases we could rename this to `ShouldSkipIf(expr) -> bool`, though. However, this quickly starts to defeat the purpose of this helper. I think that means it's trying to do too many things and should be split up.

In retrospect, the act of visiting has no business being in the helper, it's pretty hard to mess up :P The hard part is making sure we agree that a given block should be visited (the details of printing can be left to cppgen_pass). I just pushed a commit that splits this up into three separate helpers. What do you think?
- `ShouldVisitIfExpr(stmt) -> bool`
- `ShouldVisitIfBody(stmt) -> bool`
- `ShouldVisitElseBody(stmt) -> bool`